feat#17/create festival

bellringstar opened the pull request:
## ✅ PR 체크
- [ ] 하나의 PR 에는 100줄 정도의 커밋을 한다는 규칙을 지키고 있나요?
- [x] 무엇을 변경했는지 충분히 설명하고 있나요?
- [x] 새로운 기술을 사용했다면, 그 기술을 설명하고 있나요?
- [x] 이해하기 어려운 비즈니스 로직에 관해서 부연 설명을 하고 있나요?
- [x] 팀원 한 명을 Assign 했나요?
- [x] 작업 범위에 대해서 테스트를 작성하셨나요?

## 🚨 관련 이슈
closes #17 

## 🌈 작업 상황
- DB의 Festival 테이블와 연결되는 Festival 엔티티를 생성했습니다.
- Festival 생성을 위한 DTO를 생성했습니다.
- DTO에 Spring Validation을 통해 Validate를 진행하고 있습니다.
- 축제 시작 시간이 종료 시간보다 늦은 경우 예외를 발생시킵니다.
- RestDocsSupport에 Jackson 라이브러리의 LocalDateTime 파싱을 추가했습니다.
- 테스트시에 ExceptionHandler를 편하게 사용하기 위해 RestDocsSupport에 ExceptionHandler를 추가했습니다.
- ExceptionHandler의 우선순위를 조정했습니다.
- GlobalExceptionHandler에 MethodArgumentNotValidException를 처리하는 메서드를 추가했습니다.

## 📌 기타

Commit:
feat: Festival 엔티티 생성

db 테이블의 Festival 테이블에 대응되는 엔티티 클래스 작성

Commit:
feat: Festival Repository 생성

db 처리를 위한 레퍼지토리 생성

JpaRepository를 상속받아 사용한다.

Commit:
feat: Festival 생성을 위한 dto 생성

Festival 생성을 위한 dto 생성

Commit:
feat: FestivalServcie에 save 메서드 작성

축제를 생성해 db에 저장하기 위한 메서드.

requestDto를 전달받아 엔티티로 변환해 저장한 후 pk를 리턴해준다.

Commit:
feat: FestivalCreateRequestDto validation 적용

필수 값, 필드 제약조건 설정을 위한 validation 적용

시작 시간이 종료 시간보다 빠르도록 검증 로직 추가

Commit:
build: Spring validation 추가

유효성 검사를 위한 Spring Validation 의존성 추가

Commit:
feat: FestivalController 작성

Festival을 생성하기 위한 api 메서드 작성

requestDto를 받아 ApiResponse<Long>을 반환해준다. 데이터로는 festival의 pk가 나간다.

Commit:
feat: FestivalCreateRequestDto valid 테스트

Spring Validation을 통해 Validate가 정상적으로 동작하는지 확인을 하기 위한 테스트 작성

Commit:
build: Jackson의 Java 8 날짜/시간 모듈 의존성 추가

Jackson이 LocalDateTime을 파싱하지 못해 별도의 의존성을 추가해주고 설정을 추가

Commit:
fix: ApiExceptionHandler 우선순위 변경

기존 Global과 같은 Order로 설정돼있던 문제 수정

Commit:
feat: GlobalExceptionHandler 메서드 추가

MethodArgumentNotValidException를 처리하기 위한 메서드 추가

Commit:
feat: RestDocsSupport 수정

ExceptionHandler로 에외를 처리할 수 있도록 추가
LocalDateTime 파싱을 위한 설정 추가

Commit:
feat: FestivalServiceTest 작성

정상적인 dto로 festival을 생성할 수 있음을 검증하는 테스트코드 작성

Commit:
feat: FestivalControllerTest 작성

RestDocs를 위한 test 작성

minji-park-eng commented on src/main/java/com/wootecam/festivals/domain/festival/entity/Festival.java:
```diff
@@ -0,0 +1,40 @@
+    private LocalDateTime endTime;
+
+    @Builder
+    public Festival(Long organizationId, String title, String description, LocalDateTime startTime,
```
p2: private으로 변경하면 어떨까요? 생성자 대신 빌더만 사용하는 목적을 더 강제할 수 있을 것 같습니다!

bellringstar replied:
이부분은 제가 놓쳤네요! 감사합니다.

minji-park-eng commented on src/main/java/com/wootecam/festivals/domain/festival/entity/Festival.java:
```diff
@@ -0,0 +1,40 @@
+    @Column(name = "festival_id")
+    private Long id;
+    private Long organizationId; //TODO: Organization으로 ManyToOne 변경 예정
+    private String title;
+    private String description;
+    private LocalDateTime startTime;
+    private LocalDateTime endTime;
+
```
p2: @Column(nullable=false)를 추가하면 어떨까요? 코드 상에서 nullable 여부를 확인할 수 있어 필수 여부 구분이 쉬워질 것 같습니다!

bellringstar replied:
넵 코드만 봐도 알 수 있도록 하는게 좋을 것같습니다.

minji-park-eng commented on src/main/java/com/wootecam/festivals/domain/festival/dto/FestivalCreateRequestDto.java:
```diff
@@ -0,0 +1,45 @@
+import jakarta.validation.constraints.Size;
+import java.time.LocalDateTime;
+
+public record FestivalCreateRequestDto(
```
가독성 좋네요. 배워갑니다!!!

minji-park-eng commented on src/main/java/com/wootecam/festivals/domain/festival/controller/FestivalController.java:
```diff
@@ -0,0 +1,25 @@
+    private final FestivalService festivalService;
+
+    @PostMapping
+    public ApiResponse<Long> createFestival(@Valid @RequestBody FestivalCreateRequestDto requestDto) {
```
p2: 응답이 {data: 1} 식으로 올 것 같은데, 필드를 명시적으로 구분하기 위해 {data: {festivalId: 1}} 과 같이 반환되는 FestivalIdDto를 정의하면 어떨까요?

bellringstar replied:
넵 그렇게 오는거 docs 확인해서 수정하겠습니다.

minji-park-eng commented on src/main/java/com/wootecam/festivals/global/exception/ApiExceptionHandler.java:
```diff
@@ -11,7 +11,7 @@
 @RestControllerAdvice(basePackages = "com.wootecam.festivals.domain")
 @Slf4j
-@Order(Ordered.LOWEST_PRECEDENCE)
+@Order(Ordered.HIGHEST_PRECEDENCE)
```
제가...잘못 알고 있었습니다. 고쳐주셔서 감사합니다!!!

minji-park-eng commented on src/main/java/com/wootecam/festivals/global/exception/GlobalExceptionHandler.java:
```diff
@@ -40,4 +41,18 @@ public ResponseEntity<ApiErrorResponse> handleBindException(BindException except
                 .body(errorResponse);
     }
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
```
p1: MethodArgumentNotValidExceptiond이 BindException 하위라서 위에서 잡힐 것 같고, 특별히 다른 로직을 처리하는 것 같지 않습니다. 삭제하면 어떨까요?

bellringstar replied:
그렇네요.!

minji-park-eng commented on src/test/java/com/wootecam/festivals/domain/festival/controller/FestivalControllerTest.java:
```diff
@@ -0,0 +1,125 @@
+                        requestFields(
+                                fieldWithPath("organizationId").type(JsonFieldType.NULL)
+                                        .description("주최 단체 ID (필수)"),
+                                fieldWithPath("title").type(JsonFieldType.STRING)
```
P1: optional()과 fields로 필수 여부와 제약 조건을 설정할 수 있습니다! TestControllerTest에서 확인할 수 있어요!

minji-park-eng commented on src/test/java/com/wootecam/festivals/domain/festival/dto/FestivalCreateRequestDtoTest.java:
```diff
@@ -0,0 +1,58 @@
+    @ParameterizedTest
+    @MethodSource("invalidDtoProvider")
+    @DisplayName("잘못된 데이터로 DTO 생성 시 검증 실패")
+    void invalidDtoShouldFail(Long organizationId, String title, String description,
```
이렇게도 할 수 있네요!! 배워갑니다~!!

minji-park-eng reviewed (commented):
코드 잘 봤습니다! 리뷰 확인 부탁드립니다!

HyeonJun0530 reviewed (approved):
LGTM 👍

Commit:
feat: FestivalCreateResponseDto 도입에 따른 수정

FestivalCreateResponseDto를 도입해 리스폰스를 data : {festivalId : 1L} 형태로 나가도록 변경

이에 따라 기존 서비스, 컨트롤러 리턴 수정

Commit:
feat: FestivalCreateResponseDto 유효성 검증 추가

혹시나 festival이나 id가 없는데 리스폰스로 변환하려 할 경우를 방지하기 위해 유효성 검사 추가

Commit:
feat: FestivalErrorCode 작성

Festival 관련 예외코드를 설정하기 위한 클래스 작성.

Festival Data에 관한 예외 코드 추가

Commit:
feat: FestivalCreateResponseDto 도입으로인한 테스트코드 수정

기존 Long을 리턴하던 방식에서 FestivalCreateResponseDto를 리턴하도록 변경함에 따른 테스트코드 수정

Commit:
feat: Festival 엔티티 생성자 접근제어자 변경

빌더 패턴을 사용하고 있기 때문에 Pulbic으로 생성자를 열어 둘 필요가 없다. 모호성을 없애기 위해 private로 변경

Commit:
refactor: FestivalCreateResponseDto 메서드명 변경에 따른 수정

기존 toResponse에서 from으로 컨버팅 메서드 수정에 따른 코드 변경

toResponse의 경우 HttpResponse를 생성할 때 사용하는 경우가 많아 혼동의 여지가 있어 from으로 변경

Commit:
test: Festival 관련 테스트 작성을 위한 stub 작성

Festival 테스트를 작성하는데 있어 필요한 stub 작성
필요한 stub을 얻어오기 위해 정적 팩토리 메서드 추가

Commit:
test: FestivalCreateResponseDto 테스트코드 작성

변환시 문제가 있을 시 예외를 리턴하는지 검증
변환시 문제가 없을 시 정상적으로 리턴하는지를 검증

Commit:
refactor: GlobalExceptionHandler 메서드 삭제

MethodArgumentNotValidException 메서드의 경우 이미 중복으로 처리하고 있는 BindException 예외 핸들러가 있어 중복되므로 제거

Commit:
refactor: Festival 엔티티에 명시적 제약 조건 추가

- 모든 필드에 @NotNull 어노테이션 추가
- 데이터베이스 레벨 제약 조건을 위해 @Column(nullable=false) 추가
- 스키마 정의 개선을 위해 컬럼 이름과 길이 지정
- 코드 가독성 향상 및 데이터베이스 스키마와 일치
